src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

```diff
@@ -293,198 +310,144 @@ private void CheckReadArguments(byte[] buffer, int offset, int count)
 
         private async ValueTask<int> ReadAsyncCore(Memory<byte> buffer, CancellationToken cancellationToken, bool useAsync)
         {
-            // read <= count bytes from the input stream, transforming as we go.
-            // Basic idea: first we deliver any bytes we already have in the
-            // _OutputBuffer, because we know they're good.  Then, if asked to deliver
-            // more bytes, we read & transform a block at a time until either there are
-            // no bytes ready or we've delivered enough.
-            int bytesToDeliver = buffer.Length;
-            int currentOutputIndex = 0;
-            Debug.Assert(_outputBuffer != null);
-            if (_outputBufferIndex != 0)
+            while (true)
             {
-                // we have some already-transformed bytes in the output buffer
-                if (_outputBufferIndex <= buffer.Length)
+                // If there are currently any bytes stored in the output buffer, hand back as many as we can.
+                if (_outputBufferIndex != 0)
                 {
-                    _outputBuffer.AsSpan(0, _outputBufferIndex).CopyTo(buffer.Span);
-                    bytesToDeliver -= _outputBufferIndex;
-                    currentOutputIndex += _outputBufferIndex;
-                    int toClear = _outputBuffer.Length - _outputBufferIndex;
-                    CryptographicOperations.ZeroMemory(new Span<byte>(_outputBuffer, _outputBufferIndex, toClear));
-                    _outputBufferIndex = 0;
+                    int bytesToCopy = Math.Min(_outputBufferIndex, buffer.Length);
+                    if (bytesToCopy != 0)
+                    {
+                        // Copy as many bytes as we can, then shift down the remaining bytes.
+                        new ReadOnlySpan<byte>(_outputBuffer, 0, bytesToCopy).CopyTo(buffer.Span);
+                        _outputBufferIndex -= bytesToCopy;
+                        _outputBuffer.AsSpan(bytesToCopy).CopyTo(_outputBuffer);
```

Comment:
It seems like this "shift down the remaining bytes" stuff is just adding avoidable copies.

That said, since it's already there, I don't think it needs to be changed in this PR.